src/effects/backends/builtin/pitchshifteffect.cpp

```diff
@@ -0,0 +1,98 @@
+#include "effects/backends/builtin/pitchshifteffect.h"
+
+#include "util/sample.h"
+
+PitchShiftGroupState::~PitchShiftGroupState() noexcept {
+    SampleUtil::free(m_retrieveBuffer[0]);
+    SampleUtil::free(m_retrieveBuffer[1]);
+
+    VERIFY_OR_DEBUG_ASSERT(m_pRubberBand) {
+        return;
+    }
+
+    m_pRubberBand->reset();
+}
+
+// static
+QString PitchShiftEffect::getId() {
+    return "org.mixxx.effects.pitchshift";
+}
+
+//static
+EffectManifestPointer PitchShiftEffect::getManifest() {
+    EffectManifestPointer pManifest(new EffectManifest());
+
+    pManifest->setId(getId());
+    pManifest->setName(QObject::tr("Pitch Shift"));
+    pManifest->setShortName(QObject::tr("Pitch Shift"));
+    pManifest->setAuthor("The Mixxx Team");
+    pManifest->setVersion("1.0");
+    pManifest->setDescription(QObject::tr(
+            "Raises or lowers the original pitch of a sound."));
+
+    EffectManifestParameterPointer pitch = pManifest->addParameter();
+    pitch->setId("pitch");
+    pitch->setName(QObject::tr("Pitch"));
+    pitch->setShortName(QObject::tr("Pitch"));
+    pitch->setDescription(QObject::tr(
+            "The new pitch of a sound."));
+    pitch->setValueScaler(EffectManifestParameter::ValueScaler::Linear);
+    pitch->setUnitsHint(EffectManifestParameter::UnitsHint::Unknown);
+    pitch->setDefaultLinkType(EffectManifestParameter::LinkType::Linked);
+    pitch->setDefaultLinkInversion(EffectManifestParameter::LinkInversion::Inverted);
+    pitch->setNeutralPointOnScale(0.0);
+    pitch->setRange(-1.0, 0.0, 1.0);
+
+    return pManifest;
+}
+
+void PitchShiftEffect::loadEngineEffectParameters(
+        const QMap<QString, EngineEffectParameterPointer>& parameters) {
+    m_pPitchParameter = parameters.value("pitch");
+}
+
+void PitchShiftEffect::processChannel(
+        PitchShiftGroupState* pState,
+        const CSAMPLE* pInput,
+        CSAMPLE* pOutput,
+        const mixxx::EngineParameters& engineParameters,
+        const EffectEnableState enableState,
+        const GroupFeatureState& groupFeatures) {
+    Q_UNUSED(groupFeatures);
+    Q_UNUSED(enableState);
+
+    const double pitchParameter = m_pPitchParameter->value();
+
+    double pitch = 1.0;
+
+    if (pitchParameter < 0.0) {
+        pitch += pitchParameter / 2.0;
+    } else {
+        pitch += pitchParameter;
+    }
```

Comment:
Oh, interesting. Looks good to me.

Reply:
Using a lambda expression here would be overly complex and is unusal. Why not simply the ternary `?:` operator?

Reply:
I tried both and found the ternary more difficult to read to be honest. Yes, lambdas are a complex feature, but they result in very readable code when used in this way IMO.

Reply:
Tbh, I have never seen lambas used  in C++ for this purpose. We should not use uncommon patterns. If it is recommended by the C++ Core Guidelines I am ok with it, otherwise not.

Reply:
It is: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-lambda-init
Also see: https://rigtorp.se/iife/

Reply:
Nevertheless, this particular case doesn't qualify for "complex initialization" and the ternary operator should be preferred.

Reply:
Godbolt: https://compiler-explorer.com/z/Yd899MKMz

Reply:
@Swiftb0y Thanks for the comprehensive write-up!! I didn't expect the IIFE variants to be both more readable and harder to get wrong. The generated ASM doesn't really matter, compilers will get better while the code remains.

Well, I guess you forgot one version that would safe us from wasting our precious time with these discussions ;) https://compiler-explorer.com/z/dqWM64d8x
(The ASM code needs to be compared to C++/clang 14, not C++/gcc)

Reply:
> Nevertheless, this particular case doesn't qualify for "complex initialization" and the ternary operator should be preferred.

> I didn't expect the IIFE variants to be both more readable and harder to get wrong.

So for David, do we have consensus that using an IIFE makes sense here?